### PR TITLE
python3Packages.cassandra-driver: fix build by packaging new dependencies

### PR DESCRIPTION
**Pull request by risicle**
###### Motivation for this change
ZHF: #97479

This is heavily based on @turion 's #91662, but I ended up modifying it so much I thought I should just post it separately rather than give him a long list of things to fix. Though I've still left @turion as the maintainer of these - I hope that's ok @turion?

To get `NIX_REDIRECTS` working I had to add `libredirect.so` to `LD_PRELOAD`, and also add a dummy `/etc/resolv.conf` for some later tests. Also converted to `pytestCheckHook` and disabled some tests causing trouble for various reasons.

All of the packages had to be switched to the github source because the tests weren't included in any of the pypi tarballs.

`gremlinpython`'s tests do actually run if we override some version constraints and disable some sets of tests which want to connect to a running tinkerpop server.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by turion** [comment]
@risicle Thanks a lot for picking this up!

File: pkgs/development/python-modules/cassandra-driver/default.nix
Comment (turion) [inline pkgs/development/python-modules/cassandra-driver/default.nix:81]: As I don't understand all of this completely, can you add yourself as maintainer as well?

**Review by turion — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/gremlinpython/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/gremlinpython/default.nix:27]:
```suggestion
  nativeBuildInputs = [ pytestrunner ];  # simply to placate requirements
```

**Comment by risicle** [comment]
@ofborg build python37Packages.cassandra-driver

**Comment by jonringer** [comment]
getting a failure from geomet, there is a 0.2.1.post on pypi

```
  Processing ./geomet-0.2.1-py3-none-any.whl
  ERROR: Package 'geomet' requires a different Python: 3.8.6 not in '>2.6, !=3.3.*, <3.8'
cannot build derivation '/nix/store/wi330bni0g0ybm6pp0fa8ngvwd2fy78p-python3.8-cassandra-driver-3.24.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/zbppwrvd6ynb66c8dx8vm2kq9d6j0wnx-env.drv': 2 dependencies couldn't be built
[8 built (1 failed), 7 copied (1.9 MiB), 0.4 MiB DL]
error: build of '/nix/store/zbppwrvd6ynb66c8dx8vm2kq9d6j0wnx-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/100989
2 packages failed to build:
python38Packages.cassandra-driver python38Packages.geomet

5 packages built:
python27Packages.geomet python37Packages.cassandra-driver python37Packages.geomet python37Packages.gremlinpython python38Packages.gremlinpython
```

**Comment by risicle** [comment]
Hmmyes. There's no tagged release for the 3.8 fix, but I might just pull it in as a patch...

**Comment by risicle** [comment]
@ofborg build python38Packages.cassandra-driver

**Comment by SuperSandro2000** [comment]
@risicle please solve the merge conflict.

File: pkgs/development/python-modules/geomet/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/geomet/default.nix:38]:
```suggestion
  };
}
```

File: pkgs/development/python-modules/gremlinpython/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/gremlinpython/default.nix:34]:
```suggestion
  checkInputs = [ pytestCheckHook pyhamcrest radish-bdd mock ];

  # disable custom pytest report generation
  preCheck = "substituteInPlace setup.cfg --replace 'addopts' '#addopts'";

  # many tests expect a running tinkerpop server
  pytestFlagsArray = [
```

**Comment by turion** [comment]
Ping @risicle. Do you want to continue on this?

**Comment by risicle** [comment]
Sure, I just have a few things on my TODO list - currently working on a second attempt of #93083

**Comment by risicle** [comment]
Updated, tested macos 10.14, non-nixos linux.

@ofborg build python37Packages.cassandra-driver python38Packages.cassandra-driver  python39Packages.cassandra-driver

File: pkgs/development/python-modules/gremlinpython/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/gremlinpython/default.nix:33]:
```suggestion
  preCheck = ''
    substituteInPlace setup.cfg --replace 'addopts' '#addopts'
  '';
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by risicle** [comment]
I don't get it - why didn't you request these changes when you made your first pass? More to the point, how many times do we have to iterate this process until I've satisfied every last one of your personal style preferences? From the fact that I've had this on my TODO list since October, you should see I don't have a lot of time for this stuff, but you'd prefer to keep it bouncing.

**Comment by SuperSandro2000** [comment]
> I don't get it - why didn't you request these changes when you made your first pass? More to the point, how many times do we have to iterate this process until I've satisfied every last one of your personal style preferences?

I think we should be good to go after this. I am human and I can always miss something and notice it later. I think that's better than not noticing it at all. Also in the last two months I learned a lot of things and nixpkgs evolved. For example stdenv.lib got deprecated and is no longer allowed to get re-introduced.

> From the fact that I've had this on my TODO list since October, you should see I don't have a lot of time for this stuff, but you'd prefer to keep it bouncing.

I can't know if you don't have enough time ,you forgot it or just missed the notification. I let it bounce once to you which I think is reasonable to do especially after two months.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 100989` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
9 packages built:
</summary>
<ul>
<li>python37Packages.cassandra-driver</li>
<li>python37Packages.geomet</li>
<li>python37Packages.gremlinpython</li>
<li>python38Packages.cassandra-driver</li>
<li>python38Packages.geomet</li>
<li>python38Packages.gremlinpython</li>
<li>python39Packages.cassandra-driver</li>
<li>python39Packages.geomet</li>
<li>python39Packages.gremlinpython</li>
</ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 100989` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
9 packages built:
</summary>
<ul>
<li>python37Packages.cassandra-driver</li>
<li>python37Packages.geomet</li>
<li>python37Packages.gremlinpython</li>
<li>python38Packages.cassandra-driver</li>
<li>python38Packages.geomet</li>
<li>python38Packages.gremlinpython</li>
<li>python39Packages.cassandra-driver</li>
<li>python39Packages.geomet</li>
<li>python39Packages.gremlinpython</li>
</ul>
</details>

File: pkgs/development/python-modules/gremlinpython/default.nix
Comment (risicle) [reply pkgs/development/python-modules/gremlinpython/default.nix:27]: Probably possible but more fragile.

**Comment by risicle** [comment]
My point is that being overly picky about stylistic things really just puts off contributors. If something straightforward (like perhaps deciding to upstream one of your team's fixes after work for the good of the community) is going to become a multiple week saga bouncing back and forth talking about newlines and semi-colons, people just aren't going to bother because they have other things to do that feel a lot more productive.

**Comment by turion** [comment]
> My point is that being overly picky about stylistic things really just puts off contributors. If something straightforward (like perhaps deciding to upstream one of your team's fixes after work for the good of the community) is going to become a multiple week saga bouncing back and forth talking about newlines and semi-colons, people just aren't going to bother because they have other things to do that feel a lot more productive.

I agree with that. Getting a PR into nixpkgs is so much work already, even for smaller contributions. Formatting issues should be addressed by an automatic formatter. They shouldn't cost our precious human's time.

**Comment by SuperSandro2000** [comment]
> My point is that being overly picky about stylistic things really just puts off contributors.

I am not overly picky. Most of the requests are to have a somewhat uniform style to make treewide changes in the future way easier by not having to many variants of essentially the same thing.

I did not order your inputs or sort your phases. Run nixpkgs-hammer on your package and you should see what overly picky looks like.

> Getting a PR into nixpkgs is so much work already

It isn't if you follow the contributing guide. Take a look at fabaff PRs since December and a lot of them got merged without any review comments or only very small changes.

> Formatting issues should be addressed by an automatic formatter.

There is no formatter that is one stop and does everything. nixpkgs-hammering and nix-fmt should be a good option but they do not cover everything. I am not familiar with programming formatters so I can't improve the situation in a meaningful amount of time.

**Comment by turion** [comment]
> It isn't if you follow the contributing guide.

@SuperSandro2000 What exactly is the contributing guide? https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md is the only thing I know in that direction, and it doesn't give me enough information to submit PRs that can just be merged as-is. I'd like to learn more.

**Comment by SuperSandro2000** [comment]
There is also https://github.com/NixOS/nixpkgs/tree/master/doc/contributing.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 100989` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
9 packages built:
</summary>
<ul>
<li>python37Packages.cassandra-driver</li>
<li>python37Packages.geomet</li>
<li>python37Packages.gremlinpython</li>
<li>python38Packages.cassandra-driver</li>
<li>python38Packages.geomet</li>
<li>python38Packages.gremlinpython</li>
<li>python39Packages.cassandra-driver</li>
<li>python39Packages.geomet</li>
<li>python39Packages.gremlinpython</li>
</ul>
</details>

**Pull request merged**
